pkg/asset/manifests/azure/cloudproviderconfig.go

```diff
@@ -0,0 +1,111 @@
+package azure
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	azureconfig "github.com/openshift/installer/pkg/asset/installconfig/azure"
+	azuretypes "github.com/openshift/installer/pkg/types/azure"
+)
+
+const (
+	rateLimitQPSDefault          = 1.0
+	rateLimitBucketDefault       = 5
+	backoffRetriesDefault        = 6
+	backoffExponentDefault       = 1.5
+	backoffDurationDefault       = 5 // in seconds
+	backoffJitterDefault         = 1.0
+	maximumLoadBalancerRuleCount = 148 // According to Azure LB rule default limit
+
+	vmTypeStandard = "standard"
+
+	loadBalancerSkuStandard = "standard"
+
+	// Master nodes are not added to standard load balancer by default.
+	defaultExcludeMasterFromStandardLB = true
+	// Outbound SNAT is enabled by default.
+	defaultDisableOutboundSNAT = false
+)
+
+var (
+	//allows to fake the session in tests
+	fakeSession *azureconfig.Session
+)
+
+func getSession() (*azureconfig.Session, error) {
+	if fakeSession == nil {
+		return azureconfig.GetSession()
+	}
+	return fakeSession, nil
+}
+
+// CloudProviderConfig generates the cloud provider config for the vSphere platform.
+func CloudProviderConfig(p *azuretypes.Platform, resourcePrefix string) (string, error) {
```

Comment:
IMO we should not leak the entire contents of Session to this function. maybe add the TenantID and SubscriptionID to this function's API.

Reply:
also the Region, which is on platform.
so 
```
CloudProviderConfig(subscriptionID string, tenantID string, region string, resourcePrefix string) (string, error)
```

in that case, I will make the `cloudProviderConfigParameters` struct public, and use that as a parameter. Agreed?

Reply:
> also the Region, which is on platform.
> CloudProviderConfig(subscriptionID string, tenantID string, region string, resourcePrefix string) (string, error)

`p *azuretypes.Platform` was fine as-is but sure if you want to go the other extreme. i'm fine with finger print mentioned above.

> in that case, I will make the cloudProviderConfigParameters struct public, and use that as a parameter. That should make it simpler. Agreed?

no

Reply:
Either make the API of the CloudProviderConfig function take the arguments

Or

Create a type CloudProviderConfig with those 4 fields and add a function that returns the stringified cloud conf

My preference is the former

CloudProviderConfigParameter is not acceptable as this is a weird middle ground and not a great API for this package

Reply:
it's not really a weird middle ground, as the introducing a parameter object is a well known pattern that I've used for years : https://refactoring.com/catalog/introduceParameterObject.html

But fine, I'll do your second suggestion

Reply:
Looking at the https://refactoring.com/catalog/introduceParameterObject.html example,

I see 2 things:

a) they are great when that parameter type can be reused which is not the case here

b) save space on the stack

I'm not convinced this is something we should be optimizing for at this stage when we know this doesn't help us in (a) and makes our package API bad.